include/language-factory.php

```diff
@@ -0,0 +1,488 @@
+	 *     @type int     $mo_id           Optional. ID of the post storing strings translations.
+	 *     @type string  $facebook        Optional. Facebook locale.
+	 *     @type string  $home_url        Optional. Home URL in this language.
+	 *     @type string  $search_url      Optional. Home URL to use in search forms.
+	 *     @type string  $host            Optional. Host corresponding to this language.
+	 *     @type string  $flag_url        Optional. URL of the flag.
+	 *     @type string  $flag            Optional. HTML markup of the flag.
+	 *     @type string  $custom_flag_url Optional. URL of the custom flag if it exists.
+	 *     @type string  $custom_flag     Optional. HTML markup of the custom flag if it exists.
+	 *     @type int     $page_on_front   Optional. ID of the page on front in this language.
+	 *     @type int     $page_for_posts  Optional. ID of the page for posts in this language.
+	 * }
+	 *
+	 * @return PLL_Language|null A language object if given data pass sanitization, null otherwise.
+	 *
+	 * @phpstan-param array{
+	 *     term_props: array{
+	 *         language: array{
+	 *             term_id: positive-int,
+	 *             term_taxonomy_id: positive-int,
+	 *             count: int<0, max>
+	 *         },
+	 *         term_language: array{
+	 *             term_id: positive-int,
+	 *             term_taxonomy_id: positive-int,
+	 *             count: int<0, max>
+	 *         }
+	 *     },
+	 *     name: non-empty-string,
+	 *     slug: non-empty-string,
+	 *     locale: non-empty-string,
+	 *     w3c: non-empty-string,
+	 *     flag_code: non-empty-string,
+	 *     term_group: int,
+	 *     is_rtl: int<0, 1>,
+	 *     mo_id?: positive-int,
+	 *     facebook?: non-empty-string,
+	 *     home_url?: non-empty-string,
+	 *     search_url?: non-empty-string,
+	 *     host?: non-empty-string,
+	 *     flag_url?: non-empty-string,
+	 *     flag?: non-empty-string,
+	 *     custom_flag_url?: non-empty-string,
+	 *     custom_flag?: non-empty-string,
+	 *     page_on_front?:positive-int,
+	 *     page_for_posts?:positive-int
+	 * } $language_data
+	 */
+	public static function create( $language_data ) {
+		if ( isset( $language_data['slug'] ) && ( defined( 'PLL_CACHE_LANGUAGES' ) && ! PLL_CACHE_LANGUAGES ) ) { // @phpstan-ignore-line
+			$language = self::create_from_transient( $language_data['slug'] );
+			if ( ! empty( $language ) ) {
+				return $language;
+			}
+		}
+
+		$language_data = self::validate_data( $language_data );
+
+		if ( empty( $language_data ) ) {
+			return null;
+		}
+
+		return new PLL_Language( $language_data );
+	}
+
+
+	/**
+	 * Returns a language object from `pll_language_list` transient based on a given slug.
+	 *
+	 * @since 3.4
+	 *
+	 * @param string $slug Slug of the required language.
+	 * @return PLL_Language|null Language object if found, null otherwise.
+	 */
+	private static function create_from_transient( $slug ) {
```

Comment:
I don't think we should use the transient here, it is used in `PLL_Model::get_languages_list()`. Its logic should go in `create()` for the backward compatibility, without the filter and the transient things.
There's probably things to do with the `@phpstan-param` tag from `create()` then, to allow optional fields `term_taxonomy_id` etc, and make `term_props` optional.

Reply:
I moved this logic in a new method for readability, but `create_from_transient()` is really called from `create()`.
Still, I'll remove it.
But should I keep the language create from the transient at least? (as we discussed earlier)

Reply:
Done in a934c79. I kept `create_from_transient()` for readability as I said. Tell me if you want to move all of it in `create()`.

Reply:
After discussing it together, I may have misunderstood this transient thing.
To sum it up, we let `PLL_Model` handle it. But `PLL_Language_Factory::create()` still have to check deprecated properties.